From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Wiz <95232096+GameCubeGBA@users.noreply.github.com>
Date: Fri, 10 Dec 2021 13:06:09 -0500
Subject: [PATCH] 3741: Set isValid to false if
 Character.isJavaIdentifierPart(ch) is false

This looks like a mistake. The reason why it does is because the loop will only run if it is valid, and once isValid is true, |= means it will never be false, when it runs, meaning the check does nothing.

Instead, let's iterate over the string and if a character is invalid, the entire string is not an identifier
---
 .../ghidra/app/util/cparser/CPP/DefineTable.java  | 15 ++++++++-------
 1 file changed, 8 insertions(+), 7 deletions(-)

diff --git a/Ghidra/Features/Base/src/main/java/ghidra/app/util/cparser/CPP/DefineTable.java b/Ghidra/Features/Base/src/main/java/ghidra/app/util/cparser/CPP/DefineTable.java
index 11ece4ed2c..63573e1e51 100644
--- a/Ghidra/Features/Base/src/main/java/ghidra/app/util/cparser/CPP/DefineTable.java
+++ b/Ghidra/Features/Base/src/main/java/ghidra/app/util/cparser/CPP/DefineTable.java
@@ -879,15 +879,16 @@ public class DefineTable {
 			int endParen = strValue.indexOf(')', pos + 1);
 			if (endParen != -1) {
 				String subStr = strValue.substring(pos + 1, endParen);
-				if (subStr.length() > 0) {
-					int subPos = 0;
+				if (!subStr.isEmpty()) {
 					subStr = subStr.trim();
-					boolean isValid = Character.isJavaIdentifierStart(subStr.charAt(0));
-					while (isValid && subPos < subStr.length()) {
-						char ch = subStr.charAt(subPos++);
-						isValid |= Character.isJavaIdentifierPart(ch);
+					boolean isValid = true;
+					for (int subPos = 0; subPos < subStr.length(); subPos++) {
+						if (!Character.isJavaIdentifierStart(subStr.charAt(subPos))) {
+							isValid = false;
+							break;
+						}
 					}
-					// if looks like a cast, throw it away
+					// if it looks like a cast, throw it away
 					if (isValid) {
 						strValue = strValue.substring(0, pos) + strValue.substring(endParen + 1);
 						procLen = 0;
-- 
2.45.1

